File: go/integrations/langchain.go
// Package integrations provides optional hooks for third-party frameworks.
//
// DexcostCallbackHandler is duck-typed to satisfy the
// github.com/tmc/langchaingo/callbacks.Handler interface WITHOUT importing
// langchaingo as a dependency. Users who have langchaingo in their module
// can pass this handler wherever callbacks.Handler is accepted; Go's
// structural typing takes care of the rest.
//
// All recording is wrapped in recover/defer so the handler never panics
// inside a user's LLM call path.
package integrations

import (
	"context"
	"log"
	"sync"
	"time"

	"github.com/DexwoxBusiness/dexcost-sdk/go/core"
	"github.com/DexwoxBusiness/dexcost-sdk/go/pricing"
	"github.com/shopspring/decimal"
)

// pendingRun stores information captured in HandleLLMStart that is needed
// when HandleLLMEnd fires.
type pendingRun struct {
	model   string
	prompts []string
	start   time.Time
}

// DexcostCallbackHandler tracks LLM costs for langchaingo calls.
// It implements the langchaingo callbacks.Handler interface via duck typing
// (structural match) so that no import of langchaingo is required.
type DexcostCallbackHandler struct {
	buffer  core.Buffer
	pricing *pricing.Engine

	mu       sync.Mutex
	pending  map[string]*pendingRun // keyed by goroutine-local context identity
	modelHint string               // optional default model name
}

// NewDexcostCallbackHandler creates a handler that records llm_call events
// to the given buffer with costs computed by the pricing engine.
func NewDexcostCallbackHandler(buffer core.Buffer, engine *pricing.Engine) *DexcostCallbackHandler {
	return &DexcostCallbackHandler{
		buffer:  buffer,
		pricing: engine,
		pending: make(map[string]*pendingRun),
	}
}

// WithModel sets a default model name used when the output does not
// carry model information. Returns the receiver for chaining.
func (h *DexcostCallbackHandler) WithModel(model string) *DexcostCallbackHandler {
	h.modelHint = model
	return h
}

// HandleLLMStart is called when an LLM call begins.
// prompts contains the input strings sent to the model.
func (h *DexcostCallbackHandler) HandleLLMStart(ctx context.Context, prompts []string) {
	defer func() { recover() }() //nolint:errcheck

	key := pendingKeyFromCtx(ctx)
	h.mu.Lock()
	h.pending[key] = &pendingRun{
		model:   h.modelHint,
		prompts: prompts,
		start:   time.Now().UTC(),
	}
	h.mu.Unlock()
}

// HandleLLMEnd is called when an LLM call completes.
// output is expected to be a langchaingo llms.ContentResponse or similar
// struct; we extract token usage via map/struct field access.
func (h *DexcostCallbackHandler) HandleLLMEnd(ctx context.Context, output interface{}) {
	defer func() { recover() }() //nolint:errcheck

	task := core.GetCurrentTask(ctx)
	if task == nil {
		return
	}

	key := pendingKeyFromCtx(ctx)
	h.mu.Lock()
	run, ok := h.pending[key]
	delete(h.pending, key)
	h.mu.Unlock()

	startTime := time.Now().UTC()
	if ok && run != nil {
		startTime = run.start
	}

	model := h.modelHint
	if ok && run != nil && run.model != "" {
		model = run.model
	}

	inputTokens, outputTokens, cachedTokens := extractTokenUsage(output)

	latencyMs := int(time.Since(startTime).Milliseconds())

	costResult := h.pricing.GetCost(model, inputTokens, outputTokens, cachedTokens, 0)

	event := core.NewEvent(task.TaskID, core.EventTypeLLMCall)
	event.Provider = "langchaingo"
	event.Model = model
	event.CostUSD = costResult.CostUSD
	event.CostConfidence = core.CostConfidence(costResult.CostConfidence)
	event.PricingSource = core.PricingSource(costResult.PricingSource)
	event.PricingVersion = costResult.PricingVersion
	event.InputTokens = intPtr(inputTokens)
	event.OutputTokens = intPtr(outputTokens)
	if cachedTokens > 0 {
		event.CachedTokens = intPtr(cachedTokens)
	}
	event.LatencyMs = &latencyMs

	if err := h.buffer.InsertEvent(event); err != nil {
		log.Printf("[dexcost] failed to persist event: %v", err)
	}
}

// HandleLLMGenerateContentStart is called when a GenerateContent call begins.
func (h *DexcostCallbackHandler) HandleLLMGenerateContentStart(ctx context.Context, ms []interface{}) {
	defer func() { recover() }() //nolint:errcheck

	key := pendingKeyFromCtx(ctx)
	h.mu.Lock()
	h.pending[key] = &pendingRun{
		model: h.modelHint,
		start: time.Now().UTC(),
	}
	h.mu.Unlock()
}

// HandleLLMGenerateContentEnd is called when a GenerateContent call completes.
// output is expected to carry token usage information.
func (h *DexcostCallbackHandler) HandleLLMGenerateContentEnd(ctx context.Context, output interface{}) {
	defer func() { recover() }() //nolint:errcheck

	// Delegate to the same logic as HandleLLMEnd.
	h.HandleLLMEnd(ctx, output)
}

// --- No-op handlers required by callbacks.Handler ---

// HandleChainStart is a no-op.
func (h *DexcostCallbackHandler) HandleChainStart(ctx context.Context, inputs map[string]interface{}) {
}

// HandleChainEnd is a no-op.
func (h *DexcostCallbackHandler) HandleChainEnd(ctx context.Context, outputs map[string]interface{}) {
}

// HandleChainError is a no-op.
func (h *DexcostCallbackHandler) HandleChainError(ctx context.Context, err error) {
}

// HandleToolStart is a no-op.
func (h *DexcostCallbackHandler) HandleToolStart(ctx context.Context, input string) {
}

// HandleToolEnd is a no-op.
func (h *DexcostCallbackHandler) HandleToolEnd(ctx context.Context, output string) {
}

// HandleToolError is a no-op.
func (h *DexcostCallbackHandler) HandleToolError(ctx context.Context, err error) {
}

// HandleLLMError is a no-op.
func (h *DexcostCallbackHandler) HandleLLMError(ctx context.Context, err error) {
}

// HandleText is a no-op.
func (h *DexcostCallbackHandler) HandleText(ctx context.Context, text string) {
}

// HandleAgentAction is a no-op.
func (h *DexcostCallbackHandler) HandleAgentAction(ctx context.Context, action interface{}) {
}

// HandleAgentFinish is a no-op.
func (h *DexcostCallbackHandler) HandleAgentFinish(ctx context.Context, finish interface{}) {
}

// HandleRetrieverStart is a no-op.
func (h *DexcostCallbackHandler) HandleRetrieverStart(ctx context.Context, query string) {
}

// HandleRetrieverEnd is a no-op.
func (h *DexcostCallbackHandler) HandleRetrieverEnd(ctx context.Context, query string, documents []interface{}) {
}

// HandleStreamingFunc is a no-op.
func (h *DexcostCallbackHandler) HandleStreamingFunc(ctx context.Context, chunk []byte) {
}

// --- Helpers ---

// pendingKeyFromCtx derives a stable key for the pending-run map.
// We use the task ID from context so that concurrent tasks don't collide.
func pendingKeyFromCtx(ctx context.Context) string {
	task := core.GetCurrentTask(ctx)
	if task != nil {
		return task.TaskID.String()
	}
	return "unknown"
}

// extractTokenUsage attempts to pull token counts from a langchaingo output
// using map access (for map[string]interface{}) or struct field matching.
// Returns (inputTokens, outputTokens, cachedTokens).
func extractTokenUsage(output interface{}) (int, int, int) {
	if output == nil {
		return 0, 0, 0
	}

	// Try map[string]interface{} — common when output is serialised.
	if m, ok := output.(map[string]interface{}); ok {
		return intFromMap(m, "PromptTokens", "prompt_tokens", "input_tokens"),
			intFromMap(m, "CompletionTokens", "completion_tokens", "output_tokens"),
			intFromMap(m, "CachedTokens", "cached_tokens")
	}

	// Try to read exported fields via a simple interface assertion pattern.
	// langchaingo ContentResponse has a Choices field; usage may be nested.
	type usageCarrier interface {
		GetUsage() map[string]int
	}
	if uc, ok := output.(usageCarrier); ok {
		u := uc.GetUsage()
		return u["prompt_tokens"], u["completion_tokens"], u["cached_tokens"]
	}

	return 0, 0, 0
}

// intFromMap tries multiple keys and returns the first int-like value found.
func intFromMap(m map[string]interface{}, keys ...string) int {
	for _, k := range keys {
		if v, ok := m[k]; ok {
			switch val := v.(type) {
			case int:
				return val
			case int64:
				return int(val)
			case float64:
				return int(val)
			case decimal.Decimal:
				return int(val.IntPart())
			}
		}
	}
	return 0
}

// intPtr returns a pointer to the given int.
func intPtr(v int) *int {
	return &v
}

File: typescript/src/integrations/langchain.ts
/**
 * DexcostCallbackHandler — duck-typed LangChain callback handler.
 *
 * Works with any LangChain version without importing `@langchain/core`.
 * Attach to a LangChain chain/agent as a callback to automatically track
 * LLM call costs inside dexcost tasks.
 *
 * Usage:
 *   const handler = new DexcostCallbackHandler(tracker);
 *   const chain = new LLMChain({ llm, prompt, callbacks: [handler] });
 */

import { randomUUID } from "node:crypto";
import { getCurrentTask } from "../core/context.js";
import { createCostEvent } from "../core/models.js";
import type { PricingEngine } from "../pricing/engine.js";
import type { EventBuffer } from "../transport/buffer.js";
import type { CostTracker } from "../core/tracker.js";

// ---------------------------------------------------------------------------
// Internal pending-run state
// ---------------------------------------------------------------------------

interface PendingRun {
  model: string;
  startTime: number;
}

// ---------------------------------------------------------------------------
// DexcostCallbackHandler
// ---------------------------------------------------------------------------

/**
 * A duck-typed LangChain callback handler that records cost events for every
 * LLM call that occurs inside a dexcost tracked task.
 *
 * No `@langchain/core` dependency — the handler implements the subset of the
 * BaseCallbackHandler interface that LangChain calls for LLM lifecycle events.
 */
export class DexcostCallbackHandler {
  private _pricing: PricingEngine;
  private _buffer: EventBuffer;
  private _pending: Map<string, PendingRun> = new Map();

  constructor(tracker: CostTracker) {
    this._pricing = tracker.pricing;
    this._buffer = tracker.buffer;
  }

  /**
   * Called by LangChain at the start of an LLM invocation.
   *
   * @param serialized - Serialized LLM object. Model name is extracted from
   *   `serialized.kwargs.model_name` (preferred) or the last element of
   *   `serialized.id` (fallback).
   * @param prompts - The input prompt strings (unused beyond signalling start).
   * @param runId - Unique identifier for this run, used to correlate start/end.
   */
  handleLLMStart(
    serialized: Record<string, unknown>,
    _prompts: string[],
    runId: string,
  ): void {
    const model = this._extractModel(serialized);
    this._pending.set(runId, { model, startTime: Date.now() });
  }

  /**
   * Called by LangChain when an LLM invocation completes successfully.
   *
   * Records a `llm_call` cost event into the active task (if any).
   *
   * @param output - LLM output object. Token counts are read from
   *   `output.llmOutput.tokenUsage.{promptTokens, completionTokens}`.
   * @param runId - Matches the runId from the corresponding handleLLMStart.
   */
  handleLLMEnd(output: Record<string, unknown>, runId: string): void {
    const pending = this._pending.get(runId);
    if (!pending) return;

    const { model, startTime } = pending;
    const latencyMs = Date.now() - startTime;

    // Extract token counts from output.llmOutput.tokenUsage
    const llmOutput = output["llmOutput"] as Record<string, unknown> | undefined;
    const tokenUsage = llmOutput?.["tokenUsage"] as Record<string, unknown> | undefined;
    const promptTokens = (tokenUsage?.["promptTokens"] as number | undefined) ?? 0;
    const completionTokens = (tokenUsage?.["completionTokens"] as number | undefined) ?? 0;

    // Requires an active task context
    const task = getCurrentTask();
    if (!task) {
      this._pending.delete(runId);
      return;
    }

    // Compute cost
    const costResult = this._pricing.getCost(model, promptTokens, completionTokens);

    // Build event
    const event = createCostEvent({
      eventId: randomUUID(),
      taskId: task.taskId,
      eventType: "llm_call",
      provider: "langchain",
      model,
      inputTokens: promptTokens,
      outputTokens: completionTokens,
      costUsd: costResult.costUsd,
      costConfidence: costResult.costConfidence,
      pricingSource: costResult.pricingSource,
      latencyMs,
      isRetry: false,
    });

    // Persist event
    this._buffer.addEvent(event);

    // Update task aggregates
    task.llmCostUsd = task.llmCostUsd.plus(costResult.costUsd);
    task.totalCostUsd = task.totalCostUsd.plus(costResult.costUsd);
    task.totalInputTokens += promptTokens;
    task.totalOutputTokens += completionTokens;

    this._buffer.upsertTask(task);

    this._pending.delete(runId);
  }

  /**
   * Called by LangChain when an LLM invocation fails.
   *
   * Records a failure `llm_call` event (cost 0, `error_type` in details)
   * when a task context is active, mirroring the Python SDK's
   * `on_llm_error`. Never throws.
   *
   * @param error - The error that occurred.
   * @param runId - Matches the runId from the corresponding handleLLMStart.
   */
  handleLLMError(error: Error, runId: string): void {
    const pending = this._pending.get(runId);
    this._pending.delete(runId);

    const task = getCurrentTask();
    if (!task) {
      // No active task — nothing to attribute the failure to.
      return;
    }

    const model = pending?.model ?? "unknown";
    const latencyMs = pending ? Date.now() - pending.startTime : undefined;
    const errorType = error?.name ?? "Error";

    try {
      const event = createCostEvent({
        eventId: randomUUID(),
        taskId: task.taskId,
        eventType: "llm_call",
        costUsd: 0,
        costConfidence: "unknown",
        pricingSource: "unknown",
        provider: "langchain",
        model,
        latencyMs,
        isRetry: false,
        details: { error: String(error?.message ?? error), error_type: errorType },
      });
      this._buffer.addEvent(event);
      this._buffer.upsertTask(task);
    } catch {
      // dexcost errors must never crash user code
    }
  }

  // ---------------------------------------------------------------------------
  // Private helpers
  // ---------------------------------------------------------------------------

  /**
   * Extract the model identifier from a serialized LangChain LLM object.
   *
   * Preference order:
   *   1. `serialized.kwargs.model_name` — set by most ChatModel subclasses
   *   2. last element of `serialized.id` — the class name (e.g. "ChatOpenAI")
   */
  private _extractModel(serialized: Record<string, unknown>): string {
    const kwargs = serialized["kwargs"] as Record<string, unknown> | undefined;
    if (kwargs?.["model_name"] && typeof kwargs["model_name"] === "string") {
      return kwargs["model_name"];
    }

    const id = serialized["id"];
    if (Array.isArray(id) && id.length > 0) {
      const last = id[id.length - 1];
      if (typeof last === "string") return last;
    }

    return "unknown";
  }
}
